File: backend/app/api/usda.py
# backend/app/api/usda.py
"""USDA 用户路由：搜索 / 详情 / 匹配写入（原料/商品）。"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import func

from app.core.database import get_db
from app.core.security import get_current_user
from app.schemas.auth import UserResponse
from app.schemas.usda import UsdaSearchItem, UsdaFoodDetail, UsdaNutrientItem, UsdaMatchRequest
from app.models.usda import UsdaFood, UsdaFoodNutrient
from app.services.usda.index_manager import get_usda_index
from app.services.usda.matcher import match_ingredient, match_product

router = APIRouter()


@router.get("/search", response_model=list[UsdaSearchItem])
async def search_usda(
    q: str = "",
    limit: int = 50,
    db: Session = Depends(get_db),
    current_user: UserResponse = Depends(get_current_user),
):
    if not q.strip():
        return []
    index = get_usda_index()
    hits = index.search(q, limit=limit)
    if not hits:
        return []
    fdc_ids = [h["fdc_id"] for h in hits]
    score_map = {h["fdc_id"]: h["score"] for h in hits}
    rows = db.query(UsdaFood).filter(UsdaFood.fdc_id.in_(fdc_ids)).all()
    counts = dict(
        db.query(UsdaFoodNutrient.fdc_id, func.count(UsdaFoodNutrient.id))
        .filter(UsdaFoodNutrient.fdc_id.in_(fdc_ids))
        .group_by(UsdaFoodNutrient.fdc_id)
        .all()
    )
    out = [
        UsdaSearchItem(
            fdc_id=r.fdc_id,
            description=r.description,
            description_zh=r.description_zh,
            data_type=r.data_type,
            nutrient_count=counts.get(r.fdc_id, 0),
            score=score_map.get(r.fdc_id, 0),
        )
        for r in rows
    ]
    out.sort(key=lambda x: -x.score)
    return out


@router.get("/{fdc_id}", response_model=UsdaFoodDetail)
async def get_usda_food(
    fdc_id: int,
    db: Session = Depends(get_db),
    current_user: UserResponse = Depends(get_current_user),
):
    food = db.query(UsdaFood).filter(UsdaFood.fdc_id == fdc_id).first()
    if not food:
        raise HTTPException(status_code=404, detail="USDA 食材不存在")
    nutrients = (
        db.query(UsdaFoodNutrient).filter(UsdaFoodNutrient.fdc_id == fdc_id).all()
    )
    return UsdaFoodDetail(
        fdc_id=food.fdc_id,
        description=food.description,
        description_zh=food.description_zh,
        data_type=food.data_type,
        nutrients=[
            UsdaNutrientItem(
                name=n.name,
                name_zh=n.name_zh,
                amount=n.amount,
                unit_name=n.unit_name,
            )
            for n in nutrients
        ],
    )


@router.post("/match/ingredient/{ingredient_id}")
async def match_ingredient_endpoint(
    ingredient_id: int,
    body: UsdaMatchRequest,
    db: Session = Depends(get_db),
    current_user: UserResponse = Depends(get_current_user),
):
    """把 USDA 食材营养数据写入指定原料。

    TODO(权限): 现有 ingredient 写端点（ingredient_extended.py / products_entity.py）
    均只校验 `get_current_user`（登录即可），未做 owner/admin 校验。待项目引入
    统一的所有者权限依赖后，替换此处 `get_current_user`。
    """
    return match_ingredient(db, ingredient_id, body.fdc_id)


@router.post("/match/product/{product_id}")
async def match_product_endpoint(
    product_id: int,
    body: UsdaMatchRequest,
    db: Session = Depends(get_db),
    current_user: UserResponse = Depends(get_current_user),
):
    """把 USDA 食材营养数据写入指定商品的 custom_nutrition_data。

    TODO(权限): 同 match_ingredient_endpoint，待统一所有者权限依赖后替换。
    """
    return match_product(db, product_id, body.fdc_id)

File: frontend/src/api/local/handlers/usda.ts
// USDA handler — search, detail, match ingredient/product, admin download.
// Data sourced from IndexedDB stores (usda_foods, usda_food_nutrients).

import { getAll, getById, getByIndex, putOne, getDb } from '../database'

export async function searchUsda(_params: Record<string, string>, query?: any): Promise<any> {
  const q = query?.q || ''
  const limit = parseInt(query?.limit) || 50

  if (!q.trim()) return []

  const lower = q.toLowerCase()
  const all = await getAll('usda_foods')
  const matched = all.filter((f: any) => {
    const desc = (f.description || '').toLowerCase()
    const descZh = (f.description_zh || '').toLowerCase()
    return desc.includes(lower) || descZh.includes(lower)
  })

  // Sort by relevance — exact prefix match first
  matched.sort((a: any, b: any) => {
    const aDesc = (a.description || '').toLowerCase()
    const bDesc = (b.description || '').toLowerCase()
    const aPrefix = aDesc.startsWith(lower) ? 1 : 0
    const bPrefix = bDesc.startsWith(lower) ? 1 : 0
    if (aPrefix !== bPrefix) return bPrefix - aPrefix
    return (a.description || '').localeCompare(b.description || '')
  })

  const items = matched.slice(0, limit)

  // Attach nutrient counts
  const result: any[] = []
  for (const food of items) {
    const nutrients = await getByIndex('usda_food_nutrients', 'by_fdc_id', food.fdc_id)
    result.push({
      fdc_id: food.fdc_id,
      description: food.description,
      description_zh: food.description_zh || null,
      data_type: food.data_type || 'Foundation',
      nutrient_count: nutrients.length,
      score: food.description?.toLowerCase() === lower ? 100 : Math.round((lower.length / Math.max(food.description?.length || 1, 1)) * 100),
    })
  }

  return result
}

export async function getUsdaFood(params: Record<string, string>): Promise<any> {
  const fdcId = parseInt(params.fdcId)
  const food = await getById('usda_foods', fdcId)
  if (!food) throw { status: 404, message: `USDA food ${fdcId} not found` }

  const nutrients = await getByIndex('usda_food_nutrients', 'by_fdc_id', fdcId)

  return {
    fdc_id: food.fdc_id,
    description: food.description,
    description_zh: food.description_zh || null,
    data_type: food.data_type || 'Foundation',
    nutrients: nutrients.map((n: any) => ({
      name: n.name,
      name_zh: n.name_zh || null,
      amount: n.amount,
      unit_name: n.unit_name,
    })),
  }
}

export async function previewNutrition(_params: Record<string, string>, query?: any): Promise<any> {
  const fdcId = parseInt(query?.fdc_id)
  if (!fdcId) throw { status: 400, message: 'fdc_id is required' }

  const nutrients = await getByIndex('usda_food_nutrients', 'by_fdc_id', fdcId)

  // Build structured nutrition data matching backend matcher output
  const coreNutrients: Record<string, any> = {}
  const allNutrients: any[] = []
  const nutrientDetails: any[] = []

  for (const n of nutrients) {
    const item = {
      name: n.name,
      name_zh: n.name_zh || n.name,
      amount: n.amount,
      unit: n.unit_name,
    }
    allNutrients.push(item)
    nutrientDetails.push(item)

    const key = (n.name_zh || n.name).toLowerCase()
    if (!coreNutrients[key]) {
      coreNutrients[key] = { value: n.amount, unit: n.unit_name }
    }
  }

  return {
    core_nutrients: coreNutrients,
    all_nutrients: allNutrients,
    nutrient_details: nutrientDetails,
  }
}

export async function matchIngredient(params: Record<string, string>, data?: any): Promise<any> {
  const ingredientId = parseInt(params.ingredientId)
  const fdcId = data?.fdc_id
  if (!fdcId) throw { status: 400, message: 'fdc_id is required' }

  const db = await getDb()
  const tx = db.transaction('nutrition_data', 'readwrite')
  const store = tx.store

  // Delete existing nutrition data for this ingredient
  const index = store.index('by_ingredient_id')
  const existing = await index.getAll(ingredientId)
  for (const item of existing) {
    await store.delete(item.id)
  }

  // Load USDA nutrients and create nutrition_data entries
  const nutrients = await getByIndex('usda_food_nutrients', 'by_fdc_id', fdcId)
  for (const n of nutrients) {
    await store.add({
      ingredient_id: ingredientId,
      nutrient_name: n.name_zh || n.name,
      nutrient_id: n.nutrient_id || null,
      value_per_100g: n.amount,
      value: n.amount,
      unit: n.unit_name,
      source: 'usda_manual_match',
      is_verified: true,
      created_at: new Date().toISOString(),
    })
  }
  await tx.done

  return { ingredient_id: ingredientId, fdc_id: fdcId, message: 'USDA 匹配成功（本地模式）' }
}

export async function matchProduct(params: Record<string, string>, data?: any): Promise<any> {
  const productId = parseInt(params.productId)
  const fdcId = data?.fdc_id
  if (!fdcId) throw { status: 400, message: 'fdc_id is required' }

  // Load USDA nutrients
  const nutrients = await getByIndex('usda_food_nutrients', 'by_fdc_id', fdcId)
  const nutritionData = nutrients.map((n: any) => ({
    name: n.name_zh || n.name,
    value: n.amount,
    unit: n.unit_name,
  }))

  // Update product's custom_nutrition_data
  const product = await getById('products', productId)
  if (!product) throw { status: 404, message: `Product ${productId} not found` }

  await putOne('products', {
    ...product,
    id: productId,
    custom_nutrition_data: {
      nutrients: nutritionData,
      source: 'usda_manual_match',
    },
    updated_at: new Date().toISOString(),
  })

  return { product_id: productId, fdc_id: fdcId, message: 'USDA 匹配成功（本地模式）' }
}

export async function downloadUsda(_params: Record<string, string>, _data?: any): Promise<any> {
  // Local mode: mock task — actual USDA data must be pre-loaded into IndexedDB
  return { task_id: 1, message: 'USDA 下载任务已完成（本地模式使用预置数据）' }
}
